### PR TITLE
v10

### LINKED ISSUE #235: Forever retry when job timeout
hi maintainer.

I have a following job class.

```php
<?php

namespace App\Jobs;

use Illuminate\Bus\Queueable;
use Illuminate\Queue\SerializesModels;
use Illuminate\Queue\InteractsWithQueue;
use Illuminate\Contracts\Queue\ShouldQueue;
use Illuminate\Foundation\Bus\Dispatchable;

class SampleJob implements ShouldQueue
{
    use Dispatchable, InteractsWithQueue, Queueable, SerializesModels;

    public function handle()
    {
        sleep(10);
    }
}
```

When job timeout, it ignores `--tries` option and retries forever.

```console
$ while :; do php artisan queue:work --delay=1 --sleep=1 --tries=3 --timeout=3 rabbitmq; done
[2019-03-23 00:14:19][5c957a5b1b0b04.32221024] Processing: App\Jobs\SampleJob
Killed
[2019-03-23 00:14:24][5c957a5b1b0b04.32221024] Processing: App\Jobs\SampleJob
Killed
[2019-03-23 00:14:29][5c957a5b1b0b04.32221024] Processing: App\Jobs\SampleJob
Killed
[2019-03-23 00:14:35][5c957a5b1b0b04.32221024] Processing: App\Jobs\SampleJob
Killed
[2019-03-23 00:14:40][5c957a5b1b0b04.32221024] Processing: App\Jobs\SampleJob
Killed
[2019-03-23 00:14:45][5c957a5b1b0b04.32221024] Processing: App\Jobs\SampleJob
Killed
[2019-03-23 00:14:50][5c957a5b1b0b04.32221024] Processing: App\Jobs\SampleJob
Killed
[2019-03-23 00:14:55][5c957a5b1b0b04.32221024] Processing: App\Jobs\SampleJob
Killed
[2019-03-23 00:15:01][5c957a5b1b0b04.32221024] Processing: App\Jobs\SampleJob
Killed
[2019-03-23 00:15:06][5c957a5b1b0b04.32221024] Processing: App\Jobs\SampleJob
Killed
:
:
:
```

In case of redis driver,  job will fails after the number of `--tries` option.

```console
$ while :; do php artisan queue:work --delay=1 --sleep=1 --tries=3 --timeout=3 redis; done
[2019-03-23 00:17:13][dRI4jvP7vNZlMw2rnJbZgiPWf1XlyH5C] Processing: App\Jobs\SampleJob
Killed
[2019-03-23 00:18:43][dRI4jvP7vNZlMw2rnJbZgiPWf1XlyH5C] Processing: App\Jobs\SampleJob
Killed
[2019-03-23 00:20:14][dRI4jvP7vNZlMw2rnJbZgiPWf1XlyH5C] Processing: App\Jobs\SampleJob
Killed
[2019-03-23 00:21:43][dRI4jvP7vNZlMw2rnJbZgiPWf1XlyH5C] Processing: App\Jobs\SampleJob
[2019-03-23 00:21:43][dRI4jvP7vNZlMw2rnJbZgiPWf1XlyH5C] Failed:     App\Jobs\SampleJob
```

### PR DESCRIPTION
**Pull request by vyuldashev**
Changelog:

- Switch from enqueue to php-amqplib
- Fix #235
- Add support for multiple hosts
- Added console commands for declaring exchanges and queues
- Bulk push messages using `batch_basic_publish`
- No more “sleeps”. Exception will be thrown on lost connection or if any other exception occurs and process manager should be configured properly to manage such situations.

### CODE REVIEW

File: src/Console/ExchangeDeclareCommand.php [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        }
+
+        $queue->declareExchange(
+            $this->argument('name'),
+            $this->argument('type'),
+            (bool) $this->option('durable'),
+            (bool) $this->option('auto-delete')
+        );
+
+        $this->warn('Exchange already exists.');
```
Comment (golubkovden) [inline src/Console/ExchangeDeclareCommand.php]: Опечатка: очередь создана.

File: src/Console/QueueDeclareCommand.php [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+            $this->warn('Queue already exists.');
+
+            return;
+        }
+
+        $queue->declareQueue(
+            $this->argument('name'),
+            (bool) $this->option('durable'),
+            (bool) $this->option('auto-delete')
+        );
```
Comment (golubkovden) [inline src/Console/QueueDeclareCommand.php]: Добавить `$this->info(...)`

**Comment by sanderdevbaby** [comment]
How to use different exchanges now in v10?

**Comment by vyuldashev** [comment]
@sanderdevbaby v10 is not released yet and different exchanges is not in it. You can open issue and provide your ideas.

**Pull request merged**